File: JS/RecursiveChunking-BartV1.mjs
import { HfInference } from "@huggingface/inference";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";

import * as fs from "fs";
import tokenizer from "wink-tokenizer";
import "dotenv/config";

const HF_ACCESS_TOKEN = process.env.API_TOKEN;
const inference = new HfInference(HF_ACCESS_TOKEN);

var Tokenizer = tokenizer();

const MAX_INPUT_SIZE = 1024;
const MAX_OUTPUT_SIZE = 100;
const MIN_OUTPUT_SIZE = 50;

function tokenizer_len(x) {
  return Tokenizer.tokenize(x).length;
}

async function chunkify(text) {
  if (tokenizer_len(text) <= MAX_OUTPUT_SIZE) {
    return [text];
  }
  let n_chunks = Math.ceil(tokenizer_len(text) / MAX_INPUT_SIZE);
  let chunk_size = Math.ceil(tokenizer_len(text) / n_chunks) + 200;
  const splitter = new RecursiveCharacterTextSplitter({
    chunkSize: chunk_size,
    chunkOverlap: 50,
    lengthFunction: tokenizer_len,
  });
  let chunks = await splitter.createDocuments([text]);
  return chunks.map((key) => key.pageContent);
}
async function API_call(input) {
  console.log("API Call >>>>>>>>>>>", tokenizer_len(input), "tokens");
  return inference
    .summarization({
      inputs: input,
      parameters: {
        min_length: Math.min(100, tokenizer_len(input)),
        max_length: MAX_OUTPUT_SIZE,
      },
    })
    .then((data) => {
      return data["summary_text"];
    });
}

async function summarize(text, token) {
  if (text === null) {
    return null;
  }
  text = await text;
  let chunks = await chunkify(text);
  if (chunks.length === 1) {
    let ret = await API_call(chunks[0], token);
    return ret;
  }
  let new_summary = "";
  let responses = [];
  for (let i = 0; i < chunks.length; i++) {
    let ret = await API_call(chunks[i], token);
    if (!ret) {
      return null;
    }
    responses.push(ret);
  }

  for (let j = 0; j < chunks.length; j++) {
    let ret = await responses[j];
    new_summary += ret;
  }

  return summarize(new_summary);
}

summarize(fs.readFileSync("text.txt", "utf8").toString()).then((summary) =>
  console.log("auybsa")
);

// import { getEncoding, encodingForModel } from "js-tiktoken";
// const enc = getEncoding("gpt2");

// console.log(enc.encode("hello world").length);

File: RecursiveChunking_BartV1.py
from pathlib import Path
import math, os

from dotenv import load_dotenv

from langchain.text_splitter import (
    RecursiveCharacterTextSplitter,
    CharacterTextSplitter,
)
from huggingface_hub import InferenceClient
from transformers import BartTokenizer


load_dotenv(Path(".env"))
API_TOKEN = os.getenv("HF_API_TOKEN")
headers = {"Authorization": f"Bearer {API_TOKEN}"}

client = InferenceClient()
tok = BartTokenizer.from_pretrained("facebook/bart-large")
bart_tok_len = lambda x: len(tok(x)["input_ids"])
MAX_INPUT_SIZE = 1024
MAX_OUTPUT_SIZE = 100
MIN_OUTPUT_SIZE = 50


def API_call(text):
    print("API call ->>>>>>>>>>>>>>> input length:", bart_tok_len(text))
    summary = client.summarization(
        text,
        parameters={
            "min_length": min(bart_tok_len(text), MIN_OUTPUT_SIZE),
            "max_length": MAX_OUTPUT_SIZE,
        },
    )
    # print("API response <<<<<<<<<<<<<-", summary)
    return summary


def chunkify_bart(text):
    if bart_tok_len(text) <= MAX_OUTPUT_SIZE:
        return [text]
    n_chunks = math.ceil(bart_tok_len(text) / MAX_INPUT_SIZE)
    chunk_size = math.ceil(bart_tok_len(text) / n_chunks) + 200
    print(f"{n_chunks=}, {chunk_size=}")

    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=chunk_size,
        chunk_overlap=50,
        length_function=bart_tok_len,
        is_separator_regex=False,
    )
    chunks = list(
        map(lambda page: page.page_content, text_splitter.create_documents([text]))
    )
    print("Chunks sizes are:", [bart_tok_len(c) for c in chunks])
    return chunks


def summarize(comprehension):
    chunks = chunkify_bart(comprehension)
    if len(chunks) == 1:
        return API_call(chunks[0])
    chunk_summaries = [API_call(chunk) for chunk in chunks]
    return summarize(" ".join(chunk_summaries))


print(summarize(open("text.txt").read()))
